core/src/main/java/com/scalar/db/api/AuthAdmin.java

```diff
@@ -0,0 +1,166 @@
+   */
+  void grant(String username, String namespaceName, String tableName, Privilege... privileges)
+      throws ExecutionException;
+
+  /**
+   * Grants privileges to a user for the given namespace.
+   *
+   * @param username the username
+   * @param namespaceName the namespace name
+   * @param privileges the privileges
+   * @throws ExecutionException if the user does not exist or the operation fails
+   */
+  void grant(String username, String namespaceName, Privilege... privileges)
+      throws ExecutionException;
+
+  /**
+   * Revokes privileges from a user for the given table.
+   *
+   * @param username the username
+   * @param namespaceName the namespace name of the table
+   * @param tableName the table name
+   * @param privileges the privileges
+   * @throws ExecutionException if the user does not exist or the operation fails
+   */
+  void revoke(String username, String namespaceName, String tableName, Privilege... privileges)
+      throws ExecutionException;
+
+  /**
+   * Revokes a privilege from a user for the given namespace.
+   *
+   * @param username the username
+   * @param namespaceName the namespace name
+   * @param privileges the privileges
+   * @throws ExecutionException if the user does not exist or the operation fails
+   */
+  void revoke(String username, String namespaceName, Privilege... privileges)
+      throws ExecutionException;
+
+  /**
+   * Retrieve a list of usernames.
+   *
+   * @return a list of users
+   * @throws ExecutionException if the operation fails
+   */
+  List<User> getUsers() throws ExecutionException;
+
+  /**
+   * Retrieve privileges for the given namespace for the given user.
+   *
+   * @param username the username
+   * @param namespaceName the namespace name
+   * @return a set of privileges
+   * @throws ExecutionException if the user does not exist or the operation fails
+   */
+  EnumSet<Privilege> getPrivileges(String username, String namespaceName) throws ExecutionException;
+
+  /**
+   * Retrieve privileges for the given table for the given user.
+   *
+   * @param username the username
+   * @param namespaceName the namespace name of the table
+   * @param tableName the table name
+   * @return a set of privileges
+   * @throws ExecutionException if the user does not exist or the operation fails
+   */
+  EnumSet<Privilege> getPrivileges(String username, String namespaceName, String tableName)
+      throws ExecutionException;
+
+  interface User {
+    String getUsername();
+
+    boolean isSuperuser();
+  }
+
+  /** The user options. */
+  enum UserOption {
+    /** If specified, the user is created as a superuser. */
+    SUPERUSER,
+
+    /**
+     * If specified, the user is created as a non-superuser. If neither SUPERUSER nor NO_SUPERUSER
+     * is specified, the user is created as a non-superuser.
+     */
+    NO_SUPERUSER,
+  }
+
+  /** The privileges for ScalarDB operations. */
+  enum Privilege {
```

Comment:
This seems like an union of privileges for user (admin) operations, namespace operations and table operations. For instance, GRANT is for user (admin) operations and TRUNCATE is for table operations. Another option is to separate this set of privileges into each type of operations like UserPrivilege, NamespacePrivilege and TablePrivilege. This option requires additional enum definitions, but it might beneficial in the long run to avoid wrong privilege usage (e.g, TRUNCATE for a user). What do you think?

Reply:
Let me clarify the specification to ensure we are on the same page. 

Currently, we can grant privileges to a user for either a specific namespace or a table. I think we can imagine the SQL `GRANT` statement as follows:

For table privileges:
```sql
GRANT READ, WRITE, CREATE ON [TABLE] TO [USER] WITH GRANT OPTION
```

In this case, the user has `READ`, `WRITE`, `CREATE` and `GRANT` privileges for the specified table. 

- `READ`: The user can read the table.
- `WRITE`: The user can write the table.
- `CREATE`: The user can create indexes on the table.
- `GRANT`: The user can grant privileges on the table to other users.

For namespace privileges:
```sql
GRANT READ, WRITE, CREATE ON NAMESPACE [NAMESPACE] TO [USER] WITH GRANT OPTION
```

In this case, the user has `READ`, `WRITE`, `CREATE` and `GRANT` privileges for all tables in the namespace.

- `READ`: The user can read all tables in the namespace.
- `WRITE`: The user can write all tables in the namespace.
- `CREATE`: The user can create a table in the namespace or create indexes on all tables in the namespace.
- `GRANT`: The user can grant privileges on all tables in the namespace to other users.

Given this, I believe the approach is consistent and there's no need for separate privilege types for tables, namespaces, and users. What are your thoughts?

Reply:
Thanks for clarifying the usage. I understand how to use GRANT.

GRANT privilege seems a bit different type (more metadata-like?) attribute to me from other privileges (e.g. CREATE). It's like between `UserOption` and `Privilege`? It might fit `PrivilegeOption`.

> In this case, the user has READ, WRITE, CREATE and GRANT privileges for all tables in the namespace.

Understood. So, the enum is like `TablePrivilege`. BTW, we don't need to another type of privileges to allow or prohibit operations for namespaces like `GRANT CREATE ON DATABASE`?

Reply:
Ah sorry, the explanation of `GRANT` for namespace privileges was wrong.  

The correct explanation is as follows

- `GRANT`: The user can grant/revoke privileges on all tables in the namespace to/from other users. Also the user can grant/revoke privileges on the namespace to/from other users.

So for example, if a superuser issues the following:

```sql
GRANT... ON NAMESPACE ns TO user1 WITH GRANT OPTION
```

Then `user1` can issue `GRANT... ON NAMESPACE ns to ...` for other user.

What do you think?

Reply:
One more thing, if a superuser issues the following `GRANT` command:

```sql
GRANT DROP ON NAMESPACE ns TO user1
```

Then, `user1` can drop the `ns` namespace as well as all tables in the `ns` namespace.

That sounds inconsistent to you?

Reply:
Thank you for summarizing the discussion. I've fixed the Javadoc slightly.

I think we can support something like the following grammar:
```sql
GRANT { { SELECT | INSERT | UPDATE | DELETE | CREATE | DROP | TRUNCATE | ALTER } [, ...] | ALL [ PRIVILEGES ] }
    ON { [ TABLE ] table_name [, ...] | ALL TABLES IN NAMESPACE namespace_name [, ...] }
    TO user_name [, ...] [ WITH GRANT OPTION ]
```

If a user specified `ON [TABLE] table_name` in the `GRANT` statement, we call the `AuthAdmin.grant(String username, String namespaceName, String tableName, Privilege... privileges)` method. And if a user specified `ALL TABLES IN NAMESPACE namespace_name` in the `GRANT` statement, we call the `AuthAdmin.grant(String username, String namespaceName, Privilege... privileges)` method.

Also, if a user specified `WITH GRANT OPTION` and `ON [TABLE] table_name` in the GRANT statement, the granted user can issue GRANT statements, and they can grant privileges on the specified table to other users.

And if if a user specified `WITH GRANT OPTION` and `ALL TABLES IN NAMESPACE namespace_name` in the `GRANT` statement, they can grant privileges on all tables in the specified namespace to other users.

Note that users can only grant privileges that they have.

Does that make sense?

Reply:
Yes. The syntax of the GRANT statement and the behavior looks good 👍 

> As for Privilege.GRANT, I still think it might belong to a different layer, but I don't have a so strong opinion on it. Let's proceed with the current design.

This comment only considers the internal design and implementation since I still think `WITH GRANT OPTION` is a sort of option not a normal privilege. The above GRANT statement syntax itself also seems to handle `GRANT` option differently from other privileges (e.g., `UPDATE`). I checked the PostgreSQL source code, and it doesn't handle GRANT OPTION as well as other privileges https://github.com/postgres/postgres/blob/5ad49322e50459444b201e624d5df706751c9d41/src/include/nodes/parsenodes.h#L74-L93. So, IMO, maybe `Privilege.GRANT` should be removed from the enum and handled differently, although it might be a matter of taste.

Reply:
As discussed offline, we've concluded the behavior mentioned in the following comment:
https://github.com/scalar-labs/scalardb/pull/1318#discussion_r1406941025

I've also revised the Javadoc and revered the method names in ad52e62256244a28fef27e3a21055be40bf7a05b.

@komamitsu Can you please take a quick look? Thanks.

Reply:
Thanks. Looks good!